### PR TITLE
Adds SQL support to Cloud Foundry

### LINKED ISSUE #255: Add Cloud Foundry support
Our core starter should grab `GoogleCredentials` from the right spot on a PCF cluster.

### PR DESCRIPTION
**Pull request by joaoandremartins**
Fixes #255

### CODE REVIEW

**Comment by joaoandremartins** [comment]
@snicoll @dsyer FYI

**Comment by artembilan** [comment]
Some happened after this change:
```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.autoconfigure.jdbc.DataSourceProperties]: Factory method 'cloudSqlDataSourceProperties' threw exception; nested exception is java.lang.NullPointerException
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:579)
	... 50 more
Caused by: java.lang.NullPointerException
	at org.springframework.cloud.gcp.autoconfigure.sql.GcpCloudSqlAutoConfiguration$CloudSqlDataSourcePropertiesConfiguration.setCredentialsFileProperty(GcpCloudSqlAutoConfiguration.java:234)
	at org.springframework.cloud.gcp.autoconfigure.sql.GcpCloudSqlAutoConfiguration$CloudSqlDataSourcePropertiesConfiguration.cloudSqlDataSourceProperties(GcpCloudSqlAutoConfiguration.java:205)
```
See Travis report.

**Comment by joaoandremartins** [comment]
@artembilan do you want to take a quick look at this before merging?

**Comment by artembilan** [comment]
I will, but let me know that it is really a blocker and I have to switch from other stuff which is important as well.

Thanks for understanding.

File: spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java [hunk 1]
```diff
@@ -184,10 +193,17 @@ public DataSourceProperties cloudSqlDataSourceProperties(GcpCloudSqlProperties g
...
 			}
 			else {
-				LOGGER.warn("spring.datasource.url is specified. Not using generated Cloud SQL configuration");
+				LOGGER.warn("spring.datasource.url is specified. "
+						+ "Not using generated Cloud SQL configuration");
 			}
 
-			setCredentialsProperty(gcpCloudSqlProperties, gcpProperties);
+			if (environment.getProperty(
+					GcpCloudFoundryEnvironmentPostProcessor.VCAP_SERVICES_ENVVAR) != null) {
```
Comment (artembilan) [inline spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java]: I think we have that `credentials.getEncodedKey()` to consult here.
Let's try do not pull any CF specific logic into the target configurations!

File: spring-cloud-gcp-docs/src/main/asciidoc/cloudfoundry.adoc [hunk 2]
```diff
@@ -1,15 +1,11 @@
...
-will not be able to choose among bindings and will not be activated for that service.
+Spring Cloud GCP provides support for Cloud Foundry's https://docs.pivotal.io/partners/gcp-sb/index.html[GCP Service Broker].
+Our Pub/Sub, Spanner, Storage, Stackdriver Trace and Cloud SQL MySQL and PostgreSQL starters are Cloud Foundry aware and retrieve properties like project ID, credentials, etc., that are used in auto configuration from the Cloud Foundry environment.
+
+In cases like Pub/Sub's topic and subscription, or Storage's bucket name, where those parameters are not used in auto configuration, you can fetch them using the VCAP mapping provided by Spring Boot.
+For example, to retrieve the provisioned Pub/Sub topic, you can use the `vcap.services.mypubsub.credentials.topic_name` property from the application environment.
+
+NOTE: If the same service is bound to the same application more than once, the auto configuration will not be able to choose among bindings and will not be activated for that service.
+
+NOTE: If Cloud SQL MySQL and PostgreSQL are simultaneously enabled in your Cloud Foundry app, the PostgreSQL settings will have precedence over the MySQL ones.
```
Comment (artembilan) [inline spring-cloud-gcp-docs/src/main/asciidoc/cloudfoundry.adoc]: Why can't we reject such a configuration as well? The same way like you said in the previous sentence?
Thanks

**Comment by gouthamkunduru** [comment]
google-datastore with cloud foundry works?

**Comment by dsyer** [comment]
I believe there is a service broker which can be installed in a CF instance running on GCP (see links in the issue that this PR fixes above). That’s not what this issue is about though (it explicitly says SQL). You’d have to ask the GCP team (follow the links) about Datastore, and about how to use the service broker.

**Pull request merged**